### ISSUE TITLE
Tipps & Hints for a rewrite :)

### ISSUE THREAD

**Issue by kkuez** [FINDarkside/TLD-Save-Editor#78]
Hello Fin, 
I thought about rewriting your tool in java (thats my native language) and took a look over your fantastic project.
As I dig a bit into the TLD save files and compared to your code, a it was clear here no clear json was used and b you went to decode the save file form CLZF, is that correct?
I started fiddling around a bit with CLZF and the save and could not make out any success there.
Can you give me a hint where to start the journey the best?
Kind regards!

**Comment by FINDarkside** [comment]
Yeah you're correct the save file is CLZF compressed, then it's json for some part, until `m_Dict` field is an object which values are binary data stored as number array (lol). That binary data is again CLZF compressed and the output is json. After that everything should be json except that there's many places where instead of plain json the value is a string of serialized JSON, requiring you to deserialize stuff a lot of times.

If you're familiar with typescript, I have also set up working deserialization/serialization in ts, here: [tldParser.ts](https://github.com/FINDarkside/TLD-Save-Editor-3/blob/main/src/tldSave/tldParser.ts), [ObjectParser.ts](https://github.com/FINDarkside/TLD-Save-Editor-3/blob/main/src/tldSave/ObjectParser.ts). But I'm not really planning on working on that anymore, already did the interesting part of it, just can't bother to make the UI 😄 

Let me know if you have any problems or questions!

**Comment by FINDarkside** [comment]
Oh and at least in one point the profile file was invalid json, which is why I do [this](https://github.com/FINDarkside/TLD-Save-Editor/blob/b3b854a99d10fa10191bff813938ff31df449347/The%20Long%20Dark%20Save%20Editor%202/Profile.cs#L22) when deserializing and [this](https://github.com/FINDarkside/TLD-Save-Editor/blob/b3b854a99d10fa10191bff813938ff31df449347/The%20Long%20Dark%20Save%20Editor%202/Profile.cs#L40) when serializing the file.

**Comment by kkuez** [comment]
Hi Fin, 
Wow, what an awsome and heartwarmening answer, thank you a lot :) Thats what makes devolopers great, the will to share their knowledge (...and that they cba to do the UI stuff :D)!

With your tips on hand, Ill try to make the deserialization work today and the days after and will start to implement from there on.
With TS I had contact within the company, but Im more of the backend kind, thus I doubt Ill be able to make much sense out of the project, but yepp, I already saw it and will use it as a ressource for sure :)

Thanks a lot again and if you really dont mind, I would like to contact you again when I need some council here and there?
Cheers and have a nice day :)

**Comment by kkuez** [comment]
Hi once again Fin, 
Tonight Ive been fiddling around with all that and your references already got me to a decent point.
Now Im a bit confused with those binaries. Like you, I can see those annoying number arrays. You told me those contain actually "binary data" and that those are clzf encrypted too. Thats the point where I am a bit puzzled, let me begin before all that.

I wrote a small C# Program as there were no decent decryptors for java. It deencrypts the file well and gives me all contents as bytes. Those bytes I convert to chars and receive a well readable json. That json I parse to a map and yes, under the m_Dict-Node I have several entries with arrays as numbers. Now I COULD convert them to binary representations easily, but it would do me any favor.
Whats the point to go here? How can I get a deencryptable representation?
Looking forward to your answer and cheers, 
Marcel

**Comment by FINDarkside** [comment]
> Now I COULD convert them to binary representations easily, but it would not do me any favour.

That's what you should do, convert it to binary assuming that every value in the array is one byte. After that decompress it with CLZF and it'll be json. The save format doesn't make sense for sure, but it is what it is 😄  I remember once making an experiment where I removed all of this nested serialization and ended up shaving ~80% of the save size off without losing any data.

> Thanks a lot again and if you really dont mind, I would like to contact you again when I need some council here and there?

I don't mind. You can also contact me on discord if you wish (FINDarkside#1337), but github is fine as well.

> Cheers and have a nice day :)

You too :)

E: Ok I see you solved it, but I'll send this anyway since I wrote it already.
